### PR TITLE
[Feature](bangc-ops): adapt binray ops to cnnl.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Adapt binary ops to cnnl.

## 2. Modification

Please check the modified files.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

### 3.4 Summary Analysis

All release_test/release_temp passed

### CODE REVIEW

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#ifndef KERNELS_UTILS_CNNL_HELPER_H_
+#define KERNELS_UTILS_CNNL_HELPER_H_
+
+#include <iostream>
+
+#include "core/logging.h"
+#include "mlu_op.h"
+#include "cnnl.h"
+
+#define LOGERROR(ops, message) \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:32]: 这里为什么要和 LOG(ERROR) 作出区分？这个定义会有 error 字段出现吗？

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      if (ret != CNNL_STATUS_SUCCESS) {                                        \
+        LOGERROR("CNNL_HELPER", "Internal convert tensor descriptor failed."); \
+        return MLUOP_STATUS_INTERNAL_ERROR;                                    \
+      }                                                                        \
+    } else {                                                                   \
+      _desc = NULL;                                                            \
+    }                                                                          \
+  }
+
+#define CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR_V2(desc, _desc)                  \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:75]: 看上去 v1 兼容 v2 功能。为什么？

File: bangc-ops/kernels/utils/cnnl_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#ifndef KERNELS_UTILS_INCLUDE_TYPE_CONVERT_H
+#define KERNELS_UTILS_INCLUDE_TYPE_CONVERT_H
+
+#include "cnnl_helper.h"
+
+// Enumeration convert template
+template <typename SENUM, typename DENUM>
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.cpp:29]: 不如把SRC_ENUM 打全？方便读

File: bangc-ops/kernels/utils/cnnl_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#ifndef KERNELS_UTILS_INCLUDE_TYPE_CONVERT_H
+#define KERNELS_UTILS_INCLUDE_TYPE_CONVERT_H
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.cpp:24]: cpp 文件不需要加header，而且这个header 名字也对应不上

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fix, 使用LOG(ERROR)

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 2 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h:82]: v1不兼容v2,v2内部没有定义 TensorDscriptor， 目前存在一些二进制算子在外部定义了一个TensorDescriptor数组，而不能直接用V1。

File: bangc-ops/kernels/utils/cnnl_helper.cpp (hunk 3 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.cpp]: Fix

File: bangc-ops/kernels/utils/cnnl_helper.cpp (hunk 4 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.cpp]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+#include "core/logging.h"
+#include "mlu_op.h"
+#include "cnnl.h"
+
+// Checks returned status(with propose_status)
+// and if not equal returns status(ret_status)
+#define CHECK_FUNC_RETURN(ret, propose_status, message, ret_status)  \
+  {                                                                  \
+    if (ret != propose_status) {                                     \
+      LOG(ERROR)<< "MLUOPS STATUS CHECK:"<< message;                 \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:37]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      return ret_status;                                             \
+    }                                                                \
+  }
+
+// Call CNNL API
+#define CALL_CNNL(cnnl_call_ret)                                        \
+  {                                                                     \
+    mluOpStatus_t ret = mluOpStatus_t(cnnl_call_ret);                   \
+    if (ret != MLUOP_STATUS_SUCCESS) {                                  \
+      LOG(ERROR)<< "CNNL_HELPER: Internal cnnl api call error accured.";\
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:47]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+DTYPE mluOpPointerForceConvert(STYPE ptr);
+
+// TensorDescriptor
+#define CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR(desc, _desc)                     \
+  cnnlTensorDescriptor_t _desc;                                                \
+  {                                                                            \
+    if (desc != NULL) {                                                        \
+      cnnlStatus_t ret = cnnlCreateTensorDescriptor(&_desc);                   \
+      if (ret != CNNL_STATUS_SUCCESS) {                                        \
+        LOG(ERROR)<< "CNNL_HELPER: CNNL creates tensor descriptor failed.";    \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:69]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        return MLUOP_STATUS_INTERNAL_ERROR;                                    \
+      }                                                                        \
+      ret = mluOpConvertDescriptor(desc, _desc);                               \
+      if (ret != CNNL_STATUS_SUCCESS) {                                        \
+        LOG(ERROR)<< "CNNL_HELPER: Internal convert tensor descriptor failed.";\
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:74]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 2 above)
Comment (duzekunKTH) [reply bangc-ops/kernels/utils/cnnl_helper.h:82]: 那不能用v1 v2 去区分两个宏。第一个应该叫 DEFINE_CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR。现在的 v2 才应该叫 CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    } else {                                                                   \
+      _desc = NULL;                                                            \
+    }                                                                          \
+  }
+
+#define CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR_V2(desc, _desc)                  \
+  {                                                                            \
+    cnnlStatus_t ret = cnnlCreateTensorDescriptor(&_desc);                     \
+    if (ret != CNNL_STATUS_SUCCESS) {                                          \
+      LOG(ERROR)<< "CNNL_HELPER: CNNL creates tensor descriptor failed.";      \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:86]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      return MLUOP_STATUS_INTERNAL_ERROR;                                      \
+    }                                                                          \
+    ret = mluOpConvertDescriptor(desc, _desc);                                 \
+    if (ret != CNNL_STATUS_SUCCESS) {                                          \
+      LOG(ERROR)<< "CNNL_HELPER: Internal convert tensor descriptor failed.";  \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:91]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      return MLUOP_STATUS_INTERNAL_ERROR;                                      \
+    }                                                                          \
+  }
+
+#define DESTROY_CNNL_TENSOR_DESCRIPTOR(_desc)                                  \
+  {                                                                            \
+    if (_desc != NULL) {                                                       \
+      cnnlStatus_t ret = cnnlDestroyTensorDescriptor(_desc);                   \
+      if (ret != CNNL_STATUS_SUCCESS) {                                        \
+        LOG(ERROR)<< "CNNL_HELPER: CNNL destroy tensor descriptor failed.";    \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:101]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+// Handle
+#define CREATE_AND_SET_CNNL_HANDLE(handle, _handle)                  \
+  cnnlHandle_t _handle;                                              \
+  {                                                                  \
+    if (handle != NULL) {                                            \
+      cnnlStatus_t ret = cnnlCreate(&_handle);                       \
+      if (ret != CNNL_STATUS_SUCCESS) {                              \
+        LOG(ERROR)<< "CNNL_HELPER: CNNL create handle failed.";      \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:114]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        return MLUOP_STATUS_INTERNAL_ERROR;                          \
+      }                                                              \
+      ret = mluOpConvertHandle(handle, _handle);                     \
+      if (ret != CNNL_STATUS_SUCCESS) {                              \
+        LOG(ERROR)<< "CNNL_HELPER: Internal convert handle failed."; \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:119]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      }                                                              \
+    }                                                                \
+  }
+
+#define DESTROY_CNNL_HANDLE(_handle)                                  \
+  {                                                                   \
+    if (_handle != NULL) {                                            \
+      cnnlStatus_t ret = cnnlSetQueue(_handle, nullptr);              \
+      if (ret != CNNL_STATUS_SUCCESS) {                               \
+        LOG(ERROR)<< "CNNL_HELPER: Internal set handle queue failed.";\
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:130]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        return MLUOP_STATUS_INTERNAL_ERROR;                           \
+      }                                                               \
+      ret = cnnlDestroy(_handle);                                     \
+      if (ret = CNNL_STATUS_SUCCESS) {                                \
+        LOG(ERROR)<< "CNNL_HELPER: Internal destroy handle failed.";  \
```
Comment (duzekunKTH) [inline bangc-ops/kernels/utils/cnnl_helper.h:135]: 缺空格

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 5 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 6 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 11 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 12 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 13 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 14 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/utils/cnnl_helper.h (hunk 15 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/utils/cnnl_helper.h]: Fixed

File: bangc-ops/kernels/add_n/add_n.cpp [hunk 16]
```diff
@@ -20,26 +20,94 @@
...
+mluOpStatus_t MLUOP_WIN_API mluOpGetAddNWorkspaceSize(
+    mluOpHandle_t handle, const mluOpTensorDescriptor_t input_descs[],
+    const uint32_t input_num, const mluOpTensorDescriptor_t output_desc,
+    size_t *workspace_size) {
+  PARAM_CHECK("mluOpAddN_v2", handle != NULL);
+  PARAM_CHECK("mluOpAddN_v2", input_descs != NULL);
+  PARAM_CHECK("mluOpAddN_v2", output_desc != NULL);
+  PARAM_CHECK("mluOpAddN_v2", input_num >= 2);
+  DEFINE_CREATE_AND_SET_CNNL_HANDLE(handle, _handle);
+  cnnlTensorDescriptor_t *_input_descs = (cnnlTensorDescriptor_t *)malloc(
```
Comment (PetrelYy) [inline bangc-ops/kernels/add_n/add_n.cpp:34]: _*命名方式改为 input_cnnl_desc?

File: json_parser.py [hunk 17]
```diff
@@ -4,7 +4,7 @@
 
 json_file = 'build.property'
 output = 'dependency.txt'
-modules = ["cntoolkit"]
+modules = ["cntoolkit","cnnl"]
```
Comment (PetrelYy) [inline json_parser.py:7]:
```suggestion
modules = ["cntoolkit", "cnnl"]
```

**Review by PetrelYy — APPROVED** [review]

File: bangc-ops/kernels/add_n/add_n.cpp (hunk 16 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/add_n/add_n.cpp]: Fixed as  cnnl_input_desc.

File: README.md [hunk 18]
```diff
@@ -66,12 +67,13 @@ MLU-OPS 提供了以下功能：
   git submodule update --init --recursive
   ```
 
-- 准备 CNToolkit 环境
+- 准备 CNToolkit、cnnl 环境
```
Comment (PetrelYy) [inline README.md:70]: CNNL 统一大写吧

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -7190,7 +7205,6 @@ mluOpExpand(mluOpHandle_t handle,
 // Group:Fill
 /*!
  * @brief Fills the output tensor \b output with a scale \b value.
- *
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:7193]: 多删了一行

File: docs/MLU-OPS-How-To-Use-CNNL-API.md [hunk 20]
```diff
@@ -0,0 +1,201 @@
+## MLU-OPS 如何使用CNNL基础算子
```
Comment (PetrelYy) [inline docs/MLU-OPS-How-To-Use-CNNL-API.md:1]: ### MLU-OPS 如何使用 CNNL 基础算子

File: docs/MLU-OPS-How-To-Use-CNNL-API.md [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+  // 宏 DEFINE_CREATE_AND_SET_CNNL_HANDLE 内部定义并创建 CNNL handle，
+  // 并将 MLU-OPS handle 信息转换设置到 CNNL handle
+  DEFINE_CREATE_AND_SET_CNNL_HANDLE(handle, cnnl_handle);
+
+  // 宏 DEFINE_CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR 内部定义并创建 CNNL TensorDescriptor
+  // (区别于各算子特有的descriptor，例如上述的 cnnlTranposeDescriptor，通用的TensorDescriptor通过宏操作更方便)，
+  // 并将 MLU-OPS TensorDescriptor 信息转换设置到 CNNL TensorDescriptor
+  DEFINE_CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR(input_desc, cnnl_input_desc);
+  DEFINE_CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR(workspace_dst_desc,
+                                        cnnl_workspace_dst_desc);
```
[the remark this answers is not in the document]
Comment (PetrelYy) [reply docs/MLU-OPS-How-To-Use-CNNL-API.md:29]: 有个疑问，handle,desc咱们封装了宏（定义，赋值），但对于算子定制的如 cnnlTransposeDescriptor_t，又是另一套写法，会不会很奇怪？

[the remark this answers is not in the document]
Comment (PetrelYy) [reply docs/MLU-OPS-How-To-Use-CNNL-API.md:29]: 代码注释建议修改为 1.2.3. or step123，如

1. 将  mluOpHandle_t、mluOpTensorDescriptor_t 等参数转换为 CNNL 公共类型参数
    宏 DEFINE_CREATE_AND_SET_CNNL_HANDLE、DEFINE_CREATE_...
    .....

2. 定义 cnnlTranspose_v2 cnnlTranposeDescriptor  类型入参 
3. 调用 cnnlTranspose_v2
4. Destroy cnnlTranspose_v2 cnnlTranposeDescriptor  类型入参 
5. Destroy CNNL 公共类型参数
    DESTROY_CNNL_TENSOR_DESCRIPTOR...

[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply docs/MLU-OPS-How-To-Use-CNNL-API.md:29]: Fixed

File: docs/MLU-OPS-How-To-Use-CNNL-API.md (hunk 20 above)
Comment (DanieeelLiu) [reply docs/MLU-OPS-How-To-Use-CNNL-API.md:1]: Fixed

File: bangc-ops/mlu_op.h (hunk 19 above)
Comment (DanieeelLiu) [reply bangc-ops/mlu_op.h:7193]: Fixed

File: README.md (hunk 18 above)
Comment (DanieeelLiu) [reply README.md]: Fixed

**Review by duzekunKTH — APPROVED** [review]

File: README.md [hunk 22]
```diff
@@ -66,12 +67,15 @@ MLU-OPS 提供了以下功能：
   git submodule update --init --recursive
   ```
 
-- 准备 CNToolkit 环境
+- 准备 CNToolkit、CNNL 环境
 
   ```sh
+  wget https://sdk.cambricon.com/static/Basis/MLU370_X86_ubuntu18.04/cntoolkit_x.x.x-1.ubuntuxx.xx_amd64.deb
```
Comment (Unireverse) [inline README.md:73]: 我之前的想法是  https://xxx.xxxx.com/download/sdk/version=xxx 类似这样的，而不是直接把包绝对路径给出去

File: docs/MLU-OPS-How-To-Use-CNNL-API.md [hunk 23]
```diff
@@ -0,0 +1,212 @@
+## MLU-OPS 如何使用 CNNL 基础算子
+在MLU-OPS 开发算子时，如果需要用到某个基础功能，其功能已经由CNNL 算子实现，我们不妨直接在MLU-OPS 代码中调用CNNL 的API，快速达成算子开发的目的。下面3个示例介绍了如何在MLU-OPS 代码中实现调用CNNL 算子的方法。
```
Comment (Unireverse) [inline docs/MLU-OPS-How-To-Use-CNNL-API.md:2]: 加一句，需要正确配置cnnl环境，我理解是不是需要再加上如何安装cnnl？

**Review by Unireverse — APPROVED** [review]

**Pull request merged**